src/ucp/core/ucp_worker.c

```diff
@@ -533,65 +454,56 @@ ucs_status_t ucp_worker_set_ep_failed(ucp_worker_h worker, ucp_ep_h ucp_ep,
-
     if (ucp_ep_config(ucp_ep)->key.err_mode == UCP_ERR_HANDLING_MODE_NONE) {
         /* NOTE: if user has not requested error handling on the endpoint,
          *       the failure is considered unhandled */
+        ucp_ep_update_flags(ucp_ep, UCP_EP_FLAG_FAILED, 0);
+        /* Release EP ID here to prevent protocols from sending reply */
+        ucp_ep_release_id(ucp_ep);
         ret_status = status;
         goto out;
     }
 
-    err_handle_arg = ucs_malloc(sizeof(*err_handle_arg), "ucp_worker_err_handle_arg");
-    if (err_handle_arg == NULL) {
-        ucs_error("failed to allocate ucp_worker_err_handle_arg");
-        ret_status = UCS_ERR_NO_MEMORY;
-        goto out;
-    }
-
-    err_handle_arg->ucp_ep  = ucp_ep;
-    err_handle_arg->status  = status;
-    err_handle_arg->timeout = ucs_get_time() +
-            worker->context->config.ext.err_handler_delay;
-
-    /* invoke the rest of the error handling flow from the main thread */
-    uct_worker_progress_register_safe(worker->uct,
-                                      ucp_worker_iface_err_handle_progress,
-                                      err_handle_arg, UCS_CALLBACKQ_FLAG_ONESHOT,
-                                      &prog_id);
-
-    if ((ucp_ep_ext_control(ucp_ep)->err_cb == NULL) &&
-        (ucp_ep->flags & UCP_EP_FLAG_USED)) {
-        /* do not print error if connection reset by remote peer since it can
-         * be part of user level close protocol  */
-        log_level = (status == UCS_ERR_CONNECTION_RESET) ? UCS_LOG_LEVEL_DIAG :
-                    UCS_LOG_LEVEL_ERROR;
+    /* The EP can be closed from last completion callback */
+    ucp_ep_discard_lanes(ucp_ep, status);
+    ucp_ep_reqs_purge(ucp_ep, status);
+    ucp_stream_ep_cleanup(ucp_ep);
 
-        if (lane != UCP_NULL_LANE) {
-            if (lane == ucp_ep_get_cm_lane(ucp_ep)) {
-                ucs_strncpy_safe(lane_info_str, "CM lane",
-                                 sizeof(lane_info_str));
-            } else {
-                rsc_index = ucp_ep_get_rsc_index(ucp_ep, lane);
-                tl_rsc    = &worker->context->tl_rscs[rsc_index].tl_rsc;
-
-                ucs_snprintf_safe(lane_info_str, sizeof(lane_info_str),
-                                  UCT_TL_RESOURCE_DESC_FMT,
-                                  UCT_TL_RESOURCE_DESC_ARG(tl_rsc));
-            }
+    if (ucp_ep->flags & UCP_EP_FLAG_USED) {
+        if (ucp_ep->flags & UCP_EP_FLAG_CLOSE_REQ_VALID) {
+            ucs_assert(ucp_ep->flags & UCP_EP_FLAG_CLOSED);
+            /* Promote close operation to CANCEL in case of transport error,
+             * since the disconnect event may never arrive. */
+            close_req                        = ucp_ep_ext_control(ucp_ep)->
+                close_req.req;
+            close_req->send.flush.uct_flags |= UCT_FLUSH_FLAG_CANCEL;
+            ucp_ep_local_disconnect_progress(close_req);
+        } else if (ucp_ep_ext_control(ucp_ep)->err_cb == NULL) {
```

Comment:
isn't this checked already in ucp_ep_invoke_err_cb?

Reply:
checked, but we need this to print error/debug message that error won't be handled

Reply:
need to unify the code then, for example remove the check from invoke err cb

Reply:
done